Add tutorial delete modal confirm.

jelliotartz opened the pull request:
Staging branch URL: https://tutorial-delete-modal-confirm.pfe-preview.zooniverse.org/

Fixes #4117.
  - Note: I suspect there may be other places on the project builder where delete actions do not have a modal confirm implemented. Should we close this issue when this PR is merged, and reopen when/if another is found?

Describe your changes.
  - Adds a modal confirm for the tutorial delete button in the project builder.
  - Note: modal confirm does not display for deleting individual steps of the tutorial. Happy to implement this if desired, but didn't want to assume anything so left the step deletion process as is.

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [x] Does talk load correctly?
- [x] Can you post a talk comment?

# Review Checklist

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [x] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

# Optional

- [x] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [x] If changes are made to the classifier, does the dev classifier still work?
- [x] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [x] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

eatyourgreens commented on app/partials/project-modal-editor.cjsx:
```diff
@@ -204,6 +204,17 @@ ProjectModalEditorController = React.createClass
+        <p>Are you sure you want to delete this tutorial? This action is irreversible!</p>
+        <div>
+        <button className="minor-button" autoFocus={true} onClick={resolve}>Cancel</button>
+          {' '}
```
Nice one, autofocusing the Cancel button here. 👍🏻

eatyourgreens commented:
Looks good to me. I'd suggest wrapping the reusable bits up in a helper function, so that confirmation dialogs are easier to implement elsewhere. I think it would look something like `confirm(message, successCallback, resolve)`.

jelliotartz commented:
## PR Update
I refactored `confirmProjectModalDelete` to extract `succsssCallback` and `message` on this commit. A little clarification regarding wrapping the reusable bits up in a helper function would be helpful though:
  - Are you suggesting that I make a separate file, i.e. a Confirm component, that lives somewhere else (perhaps in the `lib` folder?) and takes `message`, `successCallback`, and `resolve` as args? And then I import Confirm, define my args, and call Confirm with the args? 
  - Or, would this helper function be contained within `project-modal-editor`? 
  - Or, am I off base on both of those, and you meant something else? 

Any clarification would be helpful and appreciated :) 

[Staging branch updated](https://tutorial-delete-modal-confirm.pfe-preview.zooniverse.org/).

eatyourgreens commented:
@jelliotartz Yes, I was suggesting a separate helper function that could be imported and then called with a message and callback as arguments, so that this could be reused with any other buttons that need a confirmation step. Something like the `alert()` helper, only for confirmation dialogs instead.

jelliotartz commented:
### [Staging branch updated](https://tutorial-delete-modal-confirm.pfe-preview.zooniverse.org/).

jelliotartz commented:
Hi @eatyourgreens, would you mind taking a look at this PR when you have some time? I added a helper function for confirm, as you suggested. Thanks!

eatyourgreens commented:
Looks good to me but I thought #4117 was referring to the dustbin button on tutorial steps. @beckyrother?

beckyrother commented:
Yes, it was referring to deleting the individual steps. Glad to have it when you delete the entire tutorial though too. Can you add it to the individual steps as well please?

jelliotartz commented:
## PR Update

Added a modal confirm for step deletion. The need to pass in `index` to `handleStepRemove` made `onClick={() => resolve(successCallback())}` not work as desired - it raises a successCallback is not a function error in the console (which makes sense, as we are not expecting to pass in args to `successCallback` in the `confirm` helper function). So I added an alert in the `confirmStepRemove` function that implements the modal, so that I could adjust the `successCallback` to handle passing in `index`.

#### [Staging branch updated](https://tutorial-delete-modal-confirm.pfe-preview.zooniverse.org/).

eatyourgreens commented:
I think you can use an anonymous function to invoke that callback with an argument. I'll post an example when I'm not on my phone.

eatyourgreens commented on app/partials/project-modal-editor.cjsx:
```diff
@@ -218,6 +218,20 @@ ProjectModalEditorController = React.createClass
     @props.projectModal.update 'steps'
     @props.projectModal.save()
 
+  confirmStepRemove: (index) ->
```
You can simplify this function by writing the callback as an anonymous closure, and using the `confirm()` function.
```
confirmStepRemove: (index) ->
      message = "Are you sure you want to delete this step? This action is irreversible!"
      confirm(message, () => @handleStepRemove(index))
```

jelliotartz commented:
Thanks for the suggestion :) 
#### [Staging branch updated](https://tutorial-delete-modal-confirm.pfe-preview.zooniverse.org/).
